Userland/useradd.cpp

```diff
@@ -0,0 +1,104 @@
+#include <AK/String.h>
+#include <LibCore/CArgsParser.h>
+#include <pwd.h>
+#include <stdio.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+static constexpr uid_t BASE_UID = 1000;
+static constexpr gid_t USERS_GID = 100;
+const char* DEFAULT_SHELL = "/bin/sh";
+
+int main(int argc, char** argv)
+{
+    CArgsParser args_parser("useradd");
+    args_parser.add_arg("d", "HOME_DIR", "home directory for the new user");
+    args_parser.add_arg("u", "UID", "user ID (uid) for the new user");
+    args_parser.add_arg("g", "GROUP", "group ID (gid) for the new user");
+    args_parser.add_arg("m", "create home directory if it does not exist");
+    args_parser.add_arg("s", "SHELL", "name of default shell binary for the new user");
+    args_parser.add_arg("n", "NAME", "gecos name of the new user");
+    CArgsParserResult args = args_parser.parse(argc, argv);
+
+    if (argc <= 1) {
+        args_parser.print_usage();
+        return -1;
+    }
+
+    const char* username = argv[argc - 1]; // Always the user name!
+    uid_t uid = BASE_UID;
+    gid_t gid = USERS_GID;
+
+    // First, let's sort out the uid for the user
+    if (args.is_present("u")) {
+        bool ok;
+        uid = static_cast<uid_t>(args.get("u").to_uint(ok));
+
+        if (!ok) {
+            fprintf(stderr, "Invalid uid %s", args.get("u").characters());
+            return -1;
+        }
+    } else {
+        uid = BASE_UID;
+        for (;;) // Should we cap this range?!?!
+        {
+            if (!getpwuid(uid))
+                break;
+
+            uid++;
+        }
+    }
+
+    // Now let's sort out the gid!
+    if (args.is_present("g")) {
+        bool ok;
+        gid = static_cast<gid_t>(args.get("g").to_uint(ok));
+
+        if (!ok) {
+            fprintf(stderr, "Invalid gid %s", args.get("g").characters());
+            return -1;
+        }
+    }
+
+    auto shell = args.is_present("s") ? args.get("s").characters() : DEFAULT_SHELL;
+    auto home = args.is_present("d") ? args.get("d").characters() : "/";
+    auto gecos = args.is_present("n") ? args.get("n").characters() : "Anonymous";
+
+    if (args.is_present("m")) {
+        struct stat st;
+        char path_buf[512];
+
+        sprintf(path_buf, "/home/%s", username);
```

Comment:
This buffer can be overflowed with a long `-m` argument. Please use a `StringBuilder` instead.

Reply:
Whoops, that slipped my mind.

Reply:
Ahh, wait I remember why I did this now. I think the `UTF` conversion is being funky when passed to a function. chown gives me a `Bad Address` error.

Reply:
You can only do `builder.to_string()` once. Use `.string_view()` if you want to peek at it before building the final string (including while it's being built)

Reply:
Ahh, cheers

Reply:
I don't see where you're using UTF-8 validation (you're not supposed to, here).

But the important part is you're messing up lifetimes. `builder.to_string()` gives you a `String` that *owns* the characters. When the string gets dropped — and this will happen after this line is evaluated, since you're not storing the string to any variable — the characters get deallocated (`free()`d). So `home` is going to point to undefined garbage.

In other words, keeping a pointer to string characters is not enough to keep the string alive, you need to store the string itself somewhere too.